src/ILCompiler.WebAssembly/src/CodeGen/EvaluationStack.cs

```diff
@@ -181,18 +196,58 @@ internal abstract class StackEntry
         /// </summary>
         public TypeDesc Type { get; }
 
-        public LLVMValueRef LLVMValue { get; set; }
+        Dictionary<LLVMTypeRef, LLVMValueRef> _castValues = new Dictionary<LLVMTypeRef, LLVMValueRef>(LLVMTypeRefEqualityComparer.Instance);
+
+        public LLVMValueRef ValueAsType(LLVMTypeRef type, LLVMBuilderRef builder)
+        {
+            if (!_castValues.TryGetValue(type, out LLVMValueRef value))
```

Comment:
I'm not clear that having a cache at this level helps or that it's correct. Do we actually need it?